### ISSUE TITLE
[Tracking] Large-scale MoE-VLM training improvements

### ISSUE THREAD

**Issue by yuhezhang-ai** [NVIDIA-NeMo/Automodel#2985]
## Summary

We have been running large-scale MoE-VLM training (Qwen3.5-VL-scale, EP/TP/CP/FSDP2 mixes up to 128 GPUs) on top of Automodel and accumulated a set of fixes and features that are mature enough to upstream. This issue tracks the themed PR series; each PR ships with a CI-runnable (CPU / 1-GPU) test as its proof, with large-scale runs cited as supporting evidence.

Grouping is by coherent subsystem (one area, one reviewer, shared tests) rather than many atomic PRs, and risky/large changes are kept standalone so they cannot block the safe wins.

## PR queue

| # | Themed PR | Contents | Proof | Status | PR |
| -- | -- | -- | -- | -- | -- |
| 1 | MoE TP/EP correctness | TP-safe MoE parallelization; EP+TP gradient correctness; TE grouped-expert zero-grad handling | 264 unit tests (passing), validated in large-scale campaign | merged | NVIDIA-NeMo/Automodel#2995 |
| 2 | megatron-fsdp compatibility | compat fixes for megatron-fsdp 0.5 API (flattened DP-CP, TP DTensor local shape, API drift) | 149 unit tests (passing) | merged | NVIDIA-NeMo/Automodel#2986 |
| 3 | Vision-tower AC on the EP-MoE path | follow-up to NVIDIA-NeMo/Automodel#2840: the MoE AC path (`ep_size>1`) does not cover the vision tower; also a math-SDPA recompute `context_fn` for full-block vision AC | AC-scoping + recompute-parity unit tests (181 passing incl. regressions) | merged | NVIDIA-NeMo/Automodel#2993  <br>  <br>a relevant PR landed too: NVIDIA-NeMo/Automodel#3025 |
| 4 | CP vision sharding (frame-level) | frame-level context-parallel sharding of vision inputs; docs + 40 unit tests. Sharding scope by design: frozen ViT splits frames across CP × TP; trainable ViT splits within CP only (TP ranks recompute the ViT) to keep TP-replica gradients correct; pure-TP (cp=1) sharding not enabled | unit tests (40, passing), loss parity + throughput data from validation runs | merged | NVIDIA-NeMo/Automodel#2990 |
| 5 | Setup prewarms | fla / cuBLAS / NCCL communicator prewarms (dropped `dry_run` warmup RFC) | 19 unit tests (passing) | merged | NVIDIA-NeMo/Automodel#2992 |
| 6 | Loss & packing | memory-bounded backward for the existing chunked CE; preserve indexed packing masks under FA2; scoped AC (`non_moe` / `non_moe_no_attn`) dropped after the scope decision | 331 unit tests passing (2 skipped), including forward/gradient parity and packing-mask preservation | merged | NVIDIA-NeMo/Automodel#2996 |
| 7 | CP block-diagonal varlen attention | block-diagonal varlen attention for context parallelism (the large/risky one — intentionally standalone) | 60 parity/safety tests (passing) | merged | NVIDIA-NeMo/Automodel#2989 |
| 8 | async-TP native-linear graph shaping | emit the reshape-mm-reshape pattern Inductor's `_micro_pipeline_tp` fusion recognizes (the enable flag from [#1711](<https://github.com/NVIDIA-NeMo/Automodel/issues/1711>) exists, but `TPLinear`'s bmm path defeats the fusion) | 13 unit tests (passing) | merged | NVIDIA-NeMo/Automodel#2987 |
| 9 | Zero-numel-shard-safe fused Adam for VLM | guard against zero-numel FSDP shards hitting TE fused Adam (small dense ViT on a wide mesh); Automodel-side fix independent of any TransformerEngine timeline (separate TE issue to be filed with the deterministic repro) | 9 unit tests incl. real-DTensor empty-local-shard proof (passing) | merged | NVIDIA-NeMo/Automodel#2997 |
| 10 | Qwen3.5 MoE-VLM long-context integration | integrate the CP block-diagonal packing/GDN path from NVIDIA-NeMo/Automodel#2989 and frame-level CP vision sharding from NVIDIA-NeMo/Automodel#2990 into Qwen3.5 MoE-VLM; add the Qwen3.5-122B-A10B 128K CP+EP+packing recipe as the first large-scale target | focused integration/parity tests + [100-step 128-GPU Qwen3.5-122B 128K validation](<https://wandb.ai/Nemo-automodel/yuhez_workspace/runs/hun4pet2>) | merged | NVIDIA-NeMo/Automodel#3186 |

## Updates after the initial PR queue

Updated 2026-08-27. The table below tracks follow-on PRs linked from this issue's comments or timeline; these are additions to the original queue above.

| PR | Follow-on change | Proof | Status |
| -- | -- | -- | -- |
| NVIDIA-NeMo/Automodel#3414 | Correct FSDP + FusedLinearCrossEntropy LM-head gradients and CP unused-parameter reduction; reject unsupported packed-CP backends; improve vision frame-sharding selection | CPU regression suites, two-GPU functional tests, real Qwen3.5 CP2/FSDP2 smoke, 128-H100 Qwen3.5-122B scoped CI, and cross-domain FLCE/KD validation | merged |
| NVIDIA-NeMo/Automodel#3581 | Add a configurable local vision-shape floor for closer reduced-precision parity under high-CP frame sharding | 32-H100 ABBA parity run, two-H100 NCCL smoke, CPU/Gloo suites, GitHub CI, and scoped VLM release CI | merged |
| NVIDIA-NeMo/Automodel#3702 | Reuse packed GatedDeltaNet metadata across Qwen3.5 dense/MoE layers and activation-checkpoint recomputation | 83 focused and regression tests passed (2 skipped); real-GPU parity, TPS, and peak-memory A/B validation pending | draft |
| NVIDIA-NeMo/Automodel#3708 | Shard trainable untied input embeddings and LM heads as independent FSDP units to bound the largest reduction-buffer allocation by one table | 7 focused CPU tests passed, including real one-/two-rank Gloo FSDP parity; two-H100 CUDA/NCCL standard-CE and FLCE exact parity; exact-shape reduction group 9.473 to 4.736 GiB and peak allocated memory 23.945 to 21.630 GiB | draft |

Deferred (not in this series): native Muon optimizer integration — still under active development, will be proposed once it stabilizes.

## Already upstreamed from this effort directly

Contribution of this effort merged before this series opened:

* NVIDIA-NeMo/Automodel#2926 (@zhiqi-li) — omit unset FSDP reshard argument

## Process

Each themed PR is staged on a branch off clean `main` with its ported test run locally before opening. PRs will be linked in the table as they open. cc @yuhezhang-ai

**Comment by zhiqi-li** [comment]
## Follow-up: two untracked Qwen3.5/Qwen3.8 memory and throughput issues

I rechecked the findings from a 64-H100 Qwen3.8-27B training campaign against `main@cb03cf61` (2026-08-17) and searched both open and closed issues/PRs using the relevant symbols and failure signatures. I did not find a direct tracker or fix for the two items below.

A separate correctness issue found in the same campaign--an incomplete FSDP-sharded LM-head gradient under FusedLinearCrossEntropy--has already been fixed by [#3414](https://github.com/NVIDIA-NeMo/Automodel/pull/3414), so it is intentionally omitted here.

### 1. CP=1 packed GatedDeltaNet metadata is rebuilt inside every layer and activation-checkpoint recomputation

**Scope**

This affects the native dense and MoE Qwen3.5-family paths when all of the following hold:

- indexed sequence packing is enabled;
- `cp_size == 1`;
- the model has GatedDeltaNet/linear-attention layers;
- activation checkpointing amplifies the cost, although it is not required.

The packed-CP path already owns step-level CPU/device metadata, so this report is specifically about the non-CP packed path.

**Current behavior**

Both the [dense block](https://github.com/NVIDIA-NeMo/Automodel/blob/cb03cf61ae2da0a591516fac4f21cfc6eecf48c1/nemo_automodel/components/models/qwen3_5/model.py#L360-L376) and the [MoE block](https://github.com/NVIDIA-NeMo/Automodel/blob/cb03cf61ae2da0a591516fac4f21cfc6eecf48c1/nemo_automodel/components/models/qwen3_5_moe/model.py#L187-L217) call `get_unpad_data()` inside each linear-attention block. That helper performs two host-observable GPU scalar reads (`.item()`) and an O(number-of-documents) counting loop: [packing.py L44-L98](https://github.com/NVIDIA-NeMo/Automodel/blob/cb03cf61ae2da0a591516fac4f21cfc6eecf48c1/nemo_automodel/components/models/common/packing.py#L44-L98).

A fresh `cu_seqlens` tensor is therefore created for every layer. AutoModel then passes only the device tensor to FLA: [cp_linear_attn.py L251-L271](https://github.com/NVIDIA-NeMo/Automodel/blob/cb03cf61ae2da0a591516fac4f21cfc6eecf48c1/nemo_automodel/components/models/qwen3_5_moe/cp_linear_attn.py#L251-L271).

Current AutoModel locks both `fla-core` and `flash-linear-attention` to 0.4.2: [uv.lock L1860-L1896](https://github.com/NVIDIA-NeMo/Automodel/blob/cb03cf61ae2da0a591516fac4f21cfc6eecf48c1/uv.lock#L1860-L1896). In FLA 0.4.2:

- `prepare_chunk_indices` already accepts a CPU mirror, but without one it calls `.tolist()` on a GPU-derived tensor and then moves newly built indices back to the GPU: [index.py L111-L122](https://github.com/fla-org/flash-linear-attention/blob/v0.4.2/fla/ops/utils/index.py#L111-L122);
- its tensor cache is single-entry and compares tensor arguments by object identity: [utils.py L109-L148](https://github.com/fla-org/flash-linear-attention/blob/v0.4.2/fla/utils.py#L109-L148).

Because AutoModel creates a new device tensor in every block, the FLA cache misses every time. The result is repeated GPU-to-host synchronization plus host/device allocator churn in every GDN layer. Activation-checkpoint backward repeats the whole sequence.

For Qwen3.8-27B, which has 48 linear-attention layers, that means up to **96 metadata builds per microbatch** with full activation checkpointing, although the document boundaries are identical for the entire model forward.

**Validated patch shape**

We moved metadata construction to the text-backbone boundary, before entering checkpointed layers, and reused one immutable bundle containing:

- the indexed document mask;
- valid-token indices;
- device `cu_seqlens`;
- a CPU `cu_seqlens` mirror.

Every GDN invocation receives the same two `cu_seqlens` objects and forwards both to FLA. A focused two-layer checkpointed regression test observes four GDN invocations (forward + recompute for both layers) while asserting:

- `get_unpad_data()` is called exactly once rather than four times;
- the same device and CPU tensor objects reach all four FLA calls;
- the CPU boundary tensor is `[0, 2, 4]` for a two-document test pack.

This is a deterministic synchronization/TPS and allocator-churn defect. I am **not** claiming that it is, by itself, the sole cause of any late-step CUDA OOM; an isolated large-scale TPS A/B is still needed before attaching a percentage.

**Suggested upstream fix/tests**

1. Build packed GDN metadata once per model forward, outside activation-checkpointed blocks.
2. Pass both `cu_seqlens` and `cu_seqlens_cpu` into FLA.
3. Preserve the current block-local fallback for direct block callers.
4. Add a checkpointed unit test that asserts one metadata build and shared object identity across layers/recomputation.
5. Add dense and MoE forward/gradient parity tests for packed inputs.

### 2. The default dense FSDP strategy leaves two large untied vocabulary tables in one root FSDP unit

**Scope**

This is a peak-memory scalability issue in the generic/default dense FSDP2 path when:

- input embeddings and the LM head are untied;
- both tables are large;
- both remain in the root FSDP unit;
- gradient reduction uses fp32 (the current default).

It is not a numerical-correctness issue, and it may remain latent when enough allocator/driver headroom is available.

**Current behavior and exact allocation size**

The default mixed-precision policy sets `reduce_dtype=torch.float32`, recursively shards transformer layers, and then wraps all remaining parameters in one root unit: [parallelizer.py L424-L464](https://github.com/NVIDIA-NeMo/Automodel/blob/cb03cf61ae2da0a591516fac4f21cfc6eecf48c1/nemo_automodel/components/distributed/parallelizer.py#L424-L464). There is no separate FSDP ownership for the input embedding and output head in this path.

Qwen3.8-27B declares `tie_word_embeddings=false`, `vocab_size=248320`, and `hidden_size=5120`: [model config](https://huggingface.co/Qwen/Qwen3.8-27B/blob/main/config.json). Each table therefore contains:

```
248,320 * 5,120 = 1,271,398,400 parameters
```

When both gradients are coalesced for fp32 reduction in the root unit, the contiguous input size is:

```
2 * 1,271,398,400 * 4 bytes
  = 10,171,187,200 bytes
  = 9.473 GiB
```

In our 64-H100 reproduction, post-backward failed at step 8 requesting **9.47 GiB** while roughly 9.42 GiB was driver-visible as free. The exact equality between the requested allocation and the two-table calculation identified the root FSDP group; it was not correlated with a special or all-padding data sample.

The custom MoE FSDP path already avoids this shape: it restores any architectural tie, detects tied weights, and independently shards untied `embed_tokens` and `lm_head` before the root wrap: [moe/parallelizer.py L759-L807](https://github.com/NVIDIA-NeMo/Automodel/blob/cb03cf61ae2da0a591516fac4f21cfc6eecf48c1/nemo_automodel/components/moe/parallelizer.py#L759-L807). The generic dense strategy should use the same ownership policy.

**Validated patch shape**

Before wrapping the root model:

1. restore any configured input/output weight tie;
2. resolve `get_input_embeddings()` and `get_output_embeddings()`;
3. if the modules/weights are untied, apply `fully_shard()` to each table as a separate leaf FSDP unit;
4. if they are tied, keep the shared parameter in a single/root FSDP unit.

This changes the largest contiguous fp32 reduction input from the sum of both tables (~9.47 GiB) to one table (~4.74 GiB), without changing the reduction math.

With that split, the targeted 20-step rerun completed 20/20 and a longer run cleared the original failure point. Focused unit tests assert that:

- untied input/output tables receive distinct `fully_shard()` calls before the root;
- tied weights receive neither leaf call and remain owned exactly once by the root.

**Suggested upstream fix/tests**

The cleanest implementation may be to factor the tied/untied embedding ownership logic already present in the MoE parallelizer into a shared helper and call it from both strategies. Besides the call-structure unit tests, a small distributed test can verify gradient parity for tied and untied cases.

---

For completeness, I did **not** classify either of these campaign findings as new AutoModel bugs:

- passing `logits_to_keep=None` to skip the LM head is an integration-specific extension; AutoModel's documented FLCE contract uses `logits_to_keep=1` ([#2397](https://github.com/NVIDIA-NeMo/Automodel/pull/2397), [#2603](https://github.com/NVIDIA-NeMo/Automodel/pull/2603));
- allocator-fraction caps, per-microbatch headroom guards, and hot-path `empty_cache()` are containment policies for PyTorch/direct-CUDA allocator interaction, not demonstrated AutoModel source defects.

We have patch implementations and focused regression tests for both reported items and can split them into two reviewable PRs.

**Comment by yuhezhang-ai** [comment]
PR #3702 has been replaced by upstream-branch draft #3705: https://github.com/NVIDIA-NeMo/Automodel/pull/3705\n\nThe code commit is unchanged; the replacement corrects the accidental fork-based head branch and adds real-H100 validation:\n\n- dense and MoE packed bf16 forward/backward with activation checkpointing off/on;\n- all 140 saved outputs and parameter gradients bit-exact against the base commit;\n- metadata builds reduced from 48 to 1 without checkpointing and 96 to 1 with checkpointing in a 48-GDN-layer isolation benchmark;\n- pooled median throughput improved 61.3% / 67.6% in that deliberately narrow benchmark, with no material peak-CUDA-memory change.\n\nThe PR body includes the exact stack, methodology, and the caveat that the isolated percentage is not a projection of full 27B training throughput.
